cmd/snap-confine/apparmor-support.c

```diff
@@ -100,6 +100,24 @@ void sc_init_apparmor_support(struct sc_apparmor *apparmor)
 	} else {
 		apparmor->mode = SC_AA_INVALID;
 	}
+
+        // Check that apparmor is actually usable, i.e. that we have
+        // the right capabilities to use it (CAP_MAC_ADMIN). We test
+        // by tryingto load an empty profile. If we get a
+        // permission denied error we know we are inside an environment
+        // where apparmor is not usable, e.g. in a lxd container that
+        // did not setup apparmor stacking.
+        debug("checking aa_kernel_interface_load_policy");
+        aa_kernel_interface *ki;
+        if (aa_kernel_interface_new(&ki, NULL, NULL) < 0) {
+           die("cannot get apparmor kernel interface");
+        }
+        if (aa_kernel_interface_load_policy(ki, "x", 1) < 0) {
+           if (errno == EPERM) {
+              apparmor->mode = SC_AA_NOT_APPLICABLE;
+           }
+        }
+        aa_kernel_interface_unref(ki);
```

Comment:
I like (3) but what about a container that is run by "root" and that uses lxc.apparmor.profile=unconfined? If ithe root user inside the container has the same permissions as the user in the outside and the outside is root it sounds like in this case this test would work (i.e. we can read /sys/kernel/security/apparmor/profiles from inside the container) yet we still do not have CAP_MAC_ADMIN?

Reply:
It seems quite odd to setup `lxc.apparmor.profile=unconfined` with `security.privileged: "true"`, but MAC_ADMIN is not stripped there, so the test remains valid:
```
$ lxc exec xenial-privileged-unconfined -- cat /sys/kernel/security/apparmor/profiles|head -1
lxd-xenial-privileged-unconfined_</var/snap/lxd/common/lxd> (enforce)

$ lxc exec xenial-privileged-unconfined -- apparmor_parser -r /etc/apparmor.d/usr.sbin.tcpdump && echo PASS
PASS
```

This is quite odd for many reasons, because an apparmor policy is not setup so the policy loads in the container are actually loaded into the host, so, people are going to break in exciting ways unrelated to snapd. This isn't something snapd should be worried about-- if the user did that, there is an extremely specific reason and they are on the hook for the side-effects.